File: frontend/src/lib/reportQualifications.ts
/** Qualifications retained by every report download; mirror _report_qualifications.py. */
import type { ReportEvent } from "@/types/agent";
import { UNRESOLVED_MEANING } from "./claimCheck";

export function reportQualifications(report: ReportEvent): string[] {
  const meta = report.meta ?? {};
  const notes: string[] = [];
  if (meta.review_outcome === "unavailable") {
    notes.push("The model's editorial review was unavailable for this run.");
  } else if (meta.review_outcome === "incomplete") {
    notes.push("Review found unresolved issues in this report. See the findings below.");
  }
  const raw = meta.grounding_counts;
  const keys = ["supported", "contradicted", "unresolved", "unavailable"];
  if (typeof raw === "object" && raw !== null) {
    const counts = raw as Record<string, unknown>;
    if (keys.every(key => typeof counts[key] === "number" && Number.isInteger(counts[key]) && Number(counts[key]) >= 0)) {
      notes.push(`Automated evidence check: ${counts.supported} supported, ${counts.contradicted} possible contradictions, ${counts.unresolved} unresolved, ${counts.unavailable} not checked.`);
      notes.push(UNRESOLVED_MEANING);
    }
  }
  const groups: Array<[unknown, string]> = [
    [meta.review_notes, ""],
    [meta.unverified_sentences ?? meta.residual_deficiencies, "Unverified: "],
    [meta.untested_angles, "Not researched: "],
  ];
  for (const [values, prefix] of groups) {
    if (Array.isArray(values)) {
      notes.push(...values.filter((item): item is string => typeof item === "string" && item.trim() !== "").map(item => prefix + item));
    }
  }
  return notes;
}

File: packages/agent-server/src/disco/agent_server/_report_qualifications.py
"""Evidence and editorial qualifications carried by a stored report.

These are measurements and unresolved findings, not a second verdict. Keep
them in downloads even when a report was produced before this renderer.
"""

from disco.core import ReportEvent


def report_qualifications(report: ReportEvent) -> list[str]:
    meta = report.meta
    notes: list[str] = []
    outcome = meta.get("review_outcome")
    if outcome == "unavailable":
        notes.append("The model's editorial review was unavailable for this run.")
    elif outcome == "incomplete":
        notes.append("Review found unresolved issues in this report. See the findings below.")
    counts = meta.get("grounding_counts")
    keys = ("supported", "contradicted", "unresolved", "unavailable")
    if isinstance(counts, dict) and all(
        type(counts.get(key)) is int and counts[key] >= 0 for key in keys
    ):
        notes.append(
            f"Automated evidence check: {counts['supported']} supported, "
            f"{counts['contradicted']} possible contradictions, "
            f"{counts['unresolved']} unresolved, {counts['unavailable']} not checked."
        )
        notes.append(
            "Unresolved means the checker found no source excerpt to match the sentence "
            "against — it is not a contradiction."
        )
    unverified = meta.get("unverified_sentences")
    if unverified is None:
        unverified = meta.get("residual_deficiencies")
    groups = (
        (meta.get("review_notes"), ""),
        (unverified, "Unverified: "),
        (meta.get("untested_angles"), "Not researched: "),
    )
    for values, prefix in groups:
        if isinstance(values, list):
            notes.extend(prefix + item for item in values if isinstance(item, str) and item.strip())
    return notes
